File: nezal-ar/app/models/games/base_defender/isometric/astar.rb
module BD
  class Node
    def initialize(x,y)      
      @x=x
      @y=y
      @terrain_type =0 
      @value =0
      @h=0 
      @g=0 
      @f=0 
      @target = false
      @closed =false 
      @visited = false 
      @node= nil 
      @parent = nil
    end
    attr_accessor :x, :y, :terrain_type, :value, :h, :g, :f, :closed, :visited, :node, :parent, :target
  end
  
  class Astar
    def initialize
    end
    def get_optimal_path(map, src_node, dst_node)
      open_list = []
      open_list.push(src_node)
      while(open_list.length > 0) do
         low_ind = 0;
         0.upto(open_list.length-1) do |i|
          low_ind = i if(open_list[i].f < open_list[low_ind].f)
         end
         current_node = open_list[low_ind];
         if(current_node.x == dst_node.x && current_node.y == dst_node.y)
          curr = current_node
          ret = []
          while(!curr.parent.nil?) do
            ret.push(curr);
            curr = curr.parent;
          end
          return ret;
         end
         open_list.delete_at(low_ind);
         current_node.closed = true;
         neighbors = map.neighbors(current_node);
         0.upto(neighbors.length-1) do |i|
           neighbor = neighbors[i];
           next if(neighbor.closed)
           g_score = current_node.g + get_distance(current_node,neighbor)
           g_score_is_best = false
           if(!neighbor.visited) 
            g_score_is_best = true;
            neighbor.h = get_distance(neighbor,dst_node)
            neighbor.visited = true;
            open_list.push(neighbor);
           elsif(g_score < neighbor.g)
            g_score_is_best = true;
           end
           if(g_score_is_best) 
            neighbor.parent = current_node;
            neighbor.g = g_score;
            neighbor.f = neighbor.g + neighbor.h;
           end
          end
        end
     end
     
     def get_distance(node1,node2)
      return Math.sqrt((node1.x - node2.x) ** 2 + (node1.y - node2.y) ** 2)
     end
  end
end

File: nezal-ar/public/base-defender/js/game/isometric/astar.js
var Node = Class.create({
	x:0,
	y:0,
	terrainType :0,
	value :0,
  ignorePlace : false,
	h:0,
	g:0,
	f:0,
	closed :false,
	visited : false,
	node: null,
	parent : null,
	initialize : function(x,y){
		this.x=x
		this.y=y
	}
})
var AstarNode = Class.create({
	initialize : function(node){
		this.node = node
	}
})
var Astar = Class.create({
	Initialize : function(){
	
	},
	getOptimalPath :function(map, srcNode, dstNode){
		var openList = []
		openList.push(srcNode)
        while(openList.length > 0) {
            var lowInd = 0;
            for(var i=0; i<openList.length; i++) {
			    if(openList[i].f < openList[lowInd].f) { lowInd = i; }
		    }
		    var currentNode = openList[lowInd];
		    if(currentNode.x == dstNode.x && currentNode.y == dstNode.y) {
			    var curr = currentNode;
			    var ret = [];
			    while(curr.parent) {
				    ret.push(curr);
				    curr = curr.parent;
			    }
			    return ret;
		    }
		    openList.splice(lowInd,1);
		    currentNode.closed = true;
		    var neighbors = map.neighbors(currentNode);
		    for(var i=0; i<neighbors.length;i++) {
			    var neighbor = neighbors[i];
			    if(neighbor.closed) {
				    continue;
			    }
			    var gScore = currentNode.g + this.getDistance(currentNode,neighbor);
			    var gScoreIsBest = false;

			    if(!neighbor.visited) {
				    gScoreIsBest = true;
				    neighbor.h = this.getDistance(neighbor,dstNode)
				    neighbor.visited = true;
				    openList.push(neighbor);
			    }
			    else if(gScore < neighbor.g) {
				    gScoreIsBest = true;
			    }
			    if(gScoreIsBest) {
				    neighbor.parent = currentNode;
				    neighbor.g = gScore;
				    neighbor.f = neighbor.g + neighbor.h;
				}
		    }
        }
	},
	getDistance : function(node1,node2){
		return Math.sqrt(Math.pow((node1.x - node2.x), 2) + Math.pow((node1.y - node2.y), 2));
	}
})

function run(){
	var astar = new Astar()
	var startTime = new Date()
	var path = astar.getOptimalPath(map,map.grid[Number($('srcX').value)][Number($('srcY').value)],
	map.grid[Number($('destX').value)][Number($('destY').value)])
	
	var endTime = new Date()
	//$('time').innerHTML = (endTime - startTime) +" ms"
	if(path){
		for(var i=0;i<path.length;i++){
			//$$('.cell')[path[i].x*map.grid[0].length+path[i].y].addClassName('cellPath')
		}
	}
}
